semtypinst: fix the type instantiation cache

zerbina opened the pull request:
## Summary

Fix a severe issue with the type instantiation cache that led to the
wrong types being sometimes retrieved from the cache and multiple,
redundant instances of the same type being created. With the fix, an
important compiler-internal invariant is now true: if two `PType`s
representing nominal types have a different `itemId`, they are not the
same (in terms of type equality).

## Details

The generic type instantiation cache is used when a type application
(`tyGenericInvocation`) is evaluated (`handleGenericInvocation`). The
latter previously operated as follows:
1. the invocation as-is is looked-up in the cache. If the lookup
   succeeds (which is only possible for invocations where all arguments
   are concrete types), the cached resolved instance is returned,
   otherwise, the procedure proceeds
2. immediate type parameters used as arguments are replaced (e.g., the
   `T` for `Invoc[T]`, but not for `Invoc[Invoc[T]]`) and the
3. the modified invocation is looked-up in the cache again. On success,
   the cached resolved instance is returned. In case of no success, the
   invocation is treated as producing a not-yet-seen-before resolved
   instance

In practice, this meant that all invocations where the arguments were
not immediate `tyGenericParam`eters and concrete types always resulted
in cache misses and thus new, and possibly duplicate, instances.

Since the presence of an unfinished instance of a generic type being
present in the cache always resulted in a cache miss (the
`inst.len < key.len` check), recursive generic instantiation would lead
to infinite recursion that was only bailed out of through (mis)-use of
the `localCache` table. However, since `localCache` only remembered the
most recent `PType` associated with a generic type, and the recursion
limit counter isn't reset, this, in turn, would lead to the wrong type
being committed to the global instance cache in contrived scenarios.

### Reworked Implementation

The arguments are now fully resolved prior to cache lookup, and there's
only a single lookup instead of two. A different way to look at it is
that instantiation now happens inside-to-outside instead of outside-to-
inside.

* only inherit the flags relevant for type equality from the
  invocation -- the other flags are inherited from the resolved type-
  arguments. Inheriting all flags from the invocation like was done
  previously would cause the `tfHasMeta` being erroneously included on
  the type in case of recursive generic types
* remove the unnecessary `body.kind == tyError` check. There's an
  internal assertion at the start of `handleGenericInvocation` that
  makes the check redundant
* remove the `localCache` table; it's not needed anymore
* un-export the `searchInstTypes` procedure
* for robustness, and to prevent too deep generic type recursion, keep
  the recursion limit for type instantiations

### Removed Workarounds

Tow workarounds where necessitated by multiple `PType` instances of the
same nominal type existing:
- `sameType` had to fall back to structural type equivalence for
  `tyObject` and `tyDistinct` types that resulted from generic
  invocations (indicated by the `tfFromGeneric` flag)
- the VM's type creation logic had to canonicalize `tyObject` types so
  as to not create duplicate `VmType` instances for the same object

Both workarounds are obsolete now and thus removed.

zerbina committed:
semtypinst: fix generic type instance caching

Replace the two-pass caching process with a single-pass one. All
invocation arguments are now fully resolved prior to performing a
cache lookup, which the remaining duplicate-type issues. Another
way to look at this is that instantiation now happen from inside-
to-outside instead of outside-to-inside.

zerbina committed:
semtypinst: remove the `localCache` table

The fix also allows for removing the mis-use of `localCache` for
preventing infinite recursion (and thus `localCache` itself).

zerbina committed:
remove previously required workarounds

Most notably `object` and `distinct` type comparison can now rely purely
on ID checks, as two equal nominal types now never have a different ID
(two different types can still have the same ID).

The cache used by `vmtypegen` (which should have re-used the
global instantiation cache to begin with) is also made obsolete and can
thus be removed.

zerbina committed:
Merge branch 'devel' into fix-type-instantiation-cache-issues

zerbina committed:
test: add a regression test

saem commented on compiler/sem/semtypinst.nim:
```diff
@@ -423,72 +414,73 @@ proc instCopyType*(cl: var TReplTypeVars, t: PType): PType =
-  result.flags = header.flags
+  # make sure the 'not-nil' flag (which is essentialy a type modifier) is
+  # already present on the instance prior to caching
+  result.flags = header.flags * {tfNotNil}
```
sorry, if this is clarified elsehwere but are we assuming that the instantiated type is not nil because we are instantiating it therefore any function pointer to said type would resolve?

zerbina replied:
The `tfNotNil` flag is the only flag contributing to type equality that can currently be set on the invocation, and these flags have to be present on the instance prior to recursing, otherwise there would be issues with recursive generics.

An example of where problems would occur (`a` would get the same instance as `y`, which would be wrong):  
```nim
type Generic[T] = ref object
  x: Generic[T]

var y: Generic[int] not nil
```

zerbina replied:
In order for the code to be a bit more future-proof, I'm going to change `{tfNotNil}` to `eqTypeFlags`.

saem reviewed (approved):
LGTM

zerbina committed:
semtypinst: inherit all type-equality-relevant flags

This is a bit more future-proof, as it means that adding a new type-
equality-relevant that can be set on the invocation only requires
adjusting the `eqTypeFlags` set.

zerbina commented:
/merge
